dask/async.py

```diff
@@ -458,16 +463,28 @@ def fire_task():
             data = dict((dep, state['cache'][dep])
                         for dep in get_dependencies(dsk, key))
             # Submit
-            apply_async(execute_task, args=[key, dsk[key], data, queue,
-                                            get_id, raise_on_exception])
+            apply_async(execute_task,
+                        args=(key, marshall((dsk[key], data)),
+                              marshall, unmarshall,
+                              get_id, raise_on_exception),
+                        callback=queue.put)
+
+        def fill_pool_with_ready_tasks():
+            while state['ready'] and len(state['running']) < num_workers:
+                fire_task()
 
         # Seed initial tasks into the thread pool
-        while state['ready'] and len(state['running']) < num_workers:
-            fire_task()
+        fill_pool_with_ready_tasks()
 
         # Main loop, wait on tasks to finish, insert new ones
         while state['waiting'] or state['ready'] or state['running']:
-            key, res, tb, worker_id = queue.get()
+            key, res_info = queue.get()
+            try:
+                res, tb, worker_id = unmarshall(res_info)
+            except Exception as e:
```

Comment:
We serialize exceptions to better aid debugging on the user's tasks. If they fail, we re-raise with `remote_exception`, as done [here](https://github.com/dask/dask/pull/1653/files#diff-eda6ca8d1907e62901c8ef643dc3a223R500). Deserialization errors aren't remote errors, they occur in the main process. As such, there's no need (IMO) to catch the error, cleanup the traceback, and re-raise with `remote_exception`. However, there is a need to run the scheduler cleanup stuff. Perhaps replace this block with:

```python
            try:
                res, tb, worker_id = loads(res_info)
            except Exception as e:
                for _, _, _, _, finish in callbacks:
                    if finish:
                        finish(dsk, state, True)
                raise e
```

And then continue on [here](https://github.com/dask/dask/pull/1653/files#diff-eda6ca8d1907e62901c8ef643dc3a223R500) as before.